doc/man/man1/virgil.1
.\" Man page generated from reStructuredText.
.
.TH "VIRGIL" "1" "Aug 08, 2016" "2.0.0-beta3" "virgil-cli"
.SH NAME
virgil \- command line tool for using Virgil Security full stack functionality.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
virgil command [command opts] [command args]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
The \fBVirgil\fP program is a command line tool for using Virgil Security stack functionality:
.INDENT 0.0
.IP \(bu 2
encrypt, decrypt, sign and verify data;
.IP \(bu 2
interact with Virgil Keys Service;
.IP \(bu 2
interact with Virgil Private Keys Service.
.UNINDENT
.SH COMMON COMMANDS
.sp
\fBvirgil\-keygen(1)\fP
Generate a private key with provided parameters.
.sp
\fBvirgil\-key2pub(1)\fP
Extract a public key from the private key.
.sp
\fBvirgil\-encrypt(1)\fP
Encrypt data for given recipients who can be defined by their Virgil Keys and by passwords.
.sp
\fBvirgil\-decrypt(1)\fP
Decrypt data for a given recipient who can be defined by his public key or by his password.
.sp
\fBvirgil\-sign(1)\fP
Sign data with the private key.
.sp
\fBvirgil\-verify(1)\fP
Verify data and signature with the public key.
.sp
\fBvirgil\-exhash(1)\fP
Derives hash from the given data with PBKDF function.
.sp
\fBvirgil\-config(1)\fP
Get information about Virgil CLI configuration file.
.SH IDENTITY SERVICE COMMANDS
.sp
\fBvirgil\-identity\-confirm\-global(1)\fP
Confirmation of the Identity. Returns validation token which is required for operations with Global Virgil Cards and the confirmed Identity:
.INDENT 0.0
.IP 1. 3
\fBvirgil\-card\-create\-global(1)\fP;
.IP 2. 3
\fBvirgil\-card\-revoke\-global(1)\fP;
.IP 3. 3
\fBvirgil\-public\-key\-revoke\-global(1)\fP\&.
.UNINDENT
.sp
\fBvirgil\-identity\-confirm\-private(1)\fP
Confirmation of the Identity. Returns the validation token which is required for the operations with Private Virgil Cards and the confirmed identity:
.INDENT 0.0
.IP 1. 3
\fBvirgil\-card\-create\-private(1)\fP;
.IP 2. 3
\fBvirgil\-card\-revoke\-private(1)\fP;
.IP 3. 3
\fBvirgil\-public\-key\-revoke\-private(1)\fP\&.
.UNINDENT
.sp
\fBvirgil\-identity\-verify(1)\fP
Verify an identity Returns action id.
.sp
\fBvirgil\-identity\-valid(1)\fP
Validates the passed token. Checks whether time to live and usage count to live limits for validation token are not exceeded.
.SH KEYS SERVICE COMMANDS
.sp
\fBvirgil\-public\-key\-get(1)\fP
Get user\(aqs Virgil Public Key from the Virgil Keys service.
.sp
\fBvirgil\-public\-key\-revoke\-global(1)\fP
Revoke a group of Global Virgil Cards from the Public Keys Service connected by public\-key\-id + card\-id of one of the Cards from the group.
.sp
\fBvirgil\-public\-key\-revoke\-private(1)\fP
Revoke a group of Private Virgil Cards from the Public Keys Service connected by public\-key\-id + card\-id of one of the Cards from the group.
.SH VIRGIL CARD SERVICE COMMANDS
.sp
\fBvirgil\-card\-create\-global(1)\fP
Create a Global Virgil Card. This means \fBvirgil\-identity\-verify(1)\fP \fBvirgil\-identity\-confirm\-global(1)\fP\&.
.sp
\fBvirgil\-card\-create\-private(1)\fP
Create a Private Virgil Card. This means \fBvirgil\-identity\-confirm\-private(1)\fP\&.
.sp
\fBvirgil\-card\-search\-global(1)\fP
Search for a Global Virgil Card from the Virgil Keys Service by:
.INDENT 0.0
.IP 1. 3
application name \- search an application global Virgil Card.
.IP 2. 3
email \- search a global Virgil Card.
.UNINDENT
.sp
\fBvirgil\-card\-search\-private(1)\fP
Search the Private Virgil Card from the Virgil Keys Service.
.sp
\fBvirgil\-card\-get(1)\fP
Get user\(aqs Virgil Card from the Virgil Keys service.
.sp
\fBvirgil\-card\-revoke\-private(1)\fP
Revoke a Private Virgil Card by the card\-id.
.sp
\fBvirgil\-card\-revoke\-global(1)\fP
Revoke a Global Virgil Card by the card\-id.
.SH PRIVATE KEYS SERVICE COMMANDS
.sp
\fBvirgil\-private\-key\-add(1)\fP
Add existing the private key to the Private Keys Service.
.sp
\fBvirgil\-private\-key\-get(1)\fP
Get the private key from the Virgil Private Keys Service.
.sp
\fBvirgil\-private\-key\-del(1)\fP
Delete the private key object from the Private Keys Service.
.SH SEE ALSO
.INDENT 0.0
.IP \(bu 2
virgil\-quickstart(1)
.IP \(bu 2
virgil\-keygen(1)
.IP \(bu 2
virgil\-key2pub(1)
.IP \(bu 2
virgil\-encrypt(1)
.IP \(bu 2
virgil\-decrypt(1)
.IP \(bu 2
virgil\-sign(1)
.IP \(bu 2
virgil\-verify(1)
.IP \(bu 2
virgil\-exhash(1)
.IP \(bu 2
virgil\-config(1)
.IP \(bu 2
virgil\-identity\-confirm\-global(1)
.IP \(bu 2
virgil\-identity\-confirm\-private(1)
.IP \(bu 2
virgil\-identity\-verify(1)
.IP \(bu 2
virgil\-identity\-valid(1)
.IP \(bu 2
virgil\-public\-key\-get(1)
.IP \(bu 2
virgil\-public\-key\-revoke\-global(1)
.IP \(bu 2
virgil\-public\-key\-revoke\-private(1)
.IP \(bu 2
virgil\-card\-create\-global(1)
.IP \(bu 2
virgil\-card\-create\-private(1)
.IP \(bu 2
virgil\-card\-search\-global(1)
.IP \(bu 2
virgil\-card\-search\-private(1)
.IP \(bu 2
virgil\-card\-revoke\-private(1)
.IP \(bu 2
virgil\-card\-revoke\-global(1)
.IP \(bu 2
virgil\-private\-key\-add(1)
.IP \(bu 2
virgil\-private\-key\-get(1)
.IP \(bu 2
virgil\-private\-key\-del(1)
.UNINDENT
.SH AUTHOR
Virgil Security Inc
.SH COPYRIGHT
2016, Virgil Security Inc
.\" Generated by docutils manpage writer.
.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version v2.0.0-beta2 released 2016-07-14

## Improvement
  * [CLI-84] Add improvement to the 'virgil config'

## Changes:
  * Use virgil-sdk-cpp lib v3.2.3 based on virgil-crypto lib 1.8.1


= Version v2.0.0-beta1 released 2016-06-28

## Bugfix
  * [CLI-82] Local and global configs definition mismatch
  * [CLI-83] Encryption with the same recipient id-s in one request is possible
  * Fix name from 'key-gen' to 'keygen' in std::cerr


= Version v2.0.0-beta released 2016-06-06

## Changes:
  * Fix bugs
  * Update doc


= Version v2.0.0-rc1 released 2016-05-16

Changes:
  * Use virgil-crypto lib 1.6.0
  * Use virgil-sdk-cpp lib 3.2.0
  * Add config file


= Version 0.3.4 released 2015-08-27

Bugfix:
  * Update underlying virgil crypto library to fixed version 1.0.1.

= Version 0.3.3 released 2015-08-18

Bugfix:
  * Embed MSVC redistributable package installation to the CLI package.

Changes:
  * Add executable folder to the PATH environment variable silently on Windows.

= Version 0.3.2 released 2015-08-17

Bugfix:
  * Fix RSA and EC keys generation.

= Version 0.3.1 released 2015-08-17

Bugfix:
  * Add missing win32 runtime libraries.

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/virgil.cxx
/**
 * Copyright (C) 2016 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SPLIT_CLI

#include <iostream>
#include <string>
#include <map>
#include <fstream>
#include <iterator>

#include <cli/version.h>
#include <cli/util.h>

typedef int (*main_func)(int argc, char** argv);

// simple-action
int exhash_main(int argc, char** argv);
int keygen_main(int argc, char** argv);
int key2pub_main(int argc, char** argv);
int encrypt_main(int argc, char** argv);
int decrypt_main(int argc, char** argv);
int sign_main(int argc, char** argv);
int verify_main(int argc, char** argv);
int config_main(int argc, char** argv);

// module-action
int identity_verify_main(int argc, char** argv);
int identity_confirm_global_main(int argc, char** argv);
int identity_confirm_private_main(int argc, char** argv);
int identity_valid_main(int argc, char** argv);

int card_create_global_main(int argc, char** argv);
int card_create_private_main(int argc, char** argv);
int card_get_main(int argc, char** argv);
int card_search_private_main(int argc, char** argv);
int card_search_global_main(int argc, char** argv);
int card_revoke_global_main(int argc, char** argv);
int card_revoke_private_main(int argc, char** argv);

int public_key_get_main(int argc, char** argv);
int public_key_revoke_global_main(int argc, char** argv);
int public_key_revoke_private_main(int argc, char** argv);

int private_key_add_main(int argc, char** argv);
int private_key_get_main(int argc, char** argv);
int private_key_del_main(int argc, char** argv);

static void print_usage(std::ostream& out, const char* programName);

int main(int argc, char** argv) {
    // Parse arguments.
    if (argc < 2) {
        std::cerr << "Error: "
                  << " Required argument is missing: "
                  << "command" << std::endl;
        print_usage(std::cerr, argv[0]);
        return EXIT_FAILURE;
    }

    std::string firstArg(argv[1]);
    if (firstArg == "-h" || firstArg == "--help") {
        print_usage(std::cout, argv[0]);
        return EXIT_SUCCESS;
    } else if (firstArg == "--version") {
        cli::printVersion(std::cout, argv[0]);
        return EXIT_SUCCESS;
    }

    std::map<std::string, main_func> commandsMap;
    // simple-action
    commandsMap["exhash"] = &exhash_main;
    commandsMap["keygen"] = &keygen_main;
    commandsMap["key2pub"] = &key2pub_main;
    commandsMap["encrypt"] = &encrypt_main;
    commandsMap["decrypt"] = &decrypt_main;
    commandsMap["sign"] = &sign_main;
    commandsMap["verify"] = &verify_main;
    commandsMap["config"] = &config_main;

    // module-action
    commandsMap["identity-verify"] = &identity_verify_main;
    commandsMap["identity-confirm-global"] = &identity_confirm_global_main;
    commandsMap["identity-confirm-private"] = &identity_confirm_private_main;
    commandsMap["identity-valid"] = &identity_valid_main;

    commandsMap["card-create-global"] = &card_create_global_main;
    commandsMap["card-create-private"] = &card_create_private_main;
    commandsMap["card-get"] = &card_get_main;
    commandsMap["card-search-private"] = &card_search_private_main;
    commandsMap["card-search-global"] = &card_search_global_main;
    commandsMap["card-revoke-global"] = &card_revoke_global_main;
    commandsMap["card-revoke-private"] = &card_revoke_private_main;

    commandsMap["public-key-get"] = &public_key_get_main;
    commandsMap["public-key-revoke-global"] = &public_key_revoke_global_main;
    commandsMap["public-key-revoke-private"] = &public_key_revoke_private_main;

    commandsMap["private-key-add"] = &private_key_add_main;
    commandsMap["private-key-get"] = &private_key_get_main;
    commandsMap["private-key-del"] = &private_key_del_main;

    auto module = commandsMap.find(firstArg);
    if (module != commandsMap.end()) {
        module->second(argc - 1, argv + 1);
    } else {
        std::cerr << "Error: "
                  << "command '" << firstArg << "' not found" << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

void print_usage(std::ostream& out, const char* programName) {
    std::string doc = "command [command_opts] [command_args]\n"
                      "DESCRIPTION:\n"
                      "The 'virgil' program is a command line tool for using Virgil Security\n"
                      "stack functionality:\n"
                      "encrypt, decrypt, sign and verify data;\n"
                      "interact with Virgil Keys Service;\n"
                      "interact with Virgil Private Keys Service.\n\n"

                      "COMMON COMMANDS:\n"
                      "exhash                      Derives hash from the given data with PBKDF function.\n\n"

                      "keygen                      Generate private key with given parameters.\n\n"

                      "key2pub                     Extract Public Key from the Private Key.\n\n"

                      "encrypt                     Encrypt data for given recipients which can be\n"
                      "                            defined by Virgil Public Keys and by passwords.\n\n"

                      "decrypt                     Decrypt data for given recipient which can be\n"
                      "                            defined by Virgil Public Key or by password.\n\n"

                      "sign                        Sign data with Private Key.\n\n"

                      "verify                      Verify data with Virgil Public Key.\n\n"

                      "config                      Get information about Virgil CLI configuration file.\n\n\n"

                      "CARD COMMANDS:\n"

                      "card-create-global          Create a Global Virgil Card\n\n"

                      "card-create-private         Create a Private Virgil Card\n\n"

                      "card-revoke-global          Revoke the Global Virgil Card\n\n"

                      "card-revoke-private         Revoke the Private Virgil Card\n\n"

                      "card-search-global          Search a Global Virgil Card\n\n"

                      "card-search-private         Search a Private Virgil Card\n\n"

                      "card-get                    Get user's Virgil Card/Cards\n\n"

                      "PUBCLIC KEY COMMANDS:\n"

                      "public-key-revoke-global    Revoke a group of Global Virgil Cards\n\n"

                      "public-key-revoke-private   Revoke a group of Private Virgil Cards\n\n"

                      "public-key-get              Get Public Key\n\n\n"

                      "PRIVATE KEY COMMANDS:\n"

                      "private-key-add             Add the Private Key to Private Key Service\n\n"

                      "private-key-del             Delete the Private Key to Private Key Service\n\n"

                      "private-key-get             Get the Private Key to Private Key Service\n\n\n"

                      "IDENTITY COMMANDS:\n"

                      "identity-verify             Verify a Identity for Global Virgil Card\n\n"

                      "identity-confirm-global     Confirm the Identity for the Global Virgil Card\n\n"

                      "identity-confirm-private    Confirm the Identity for the Private Virgil Card\n\n"

                      "identity-valid              Validate the Identity for Global Virgil Card\n\n\n";

    out << "USAGE: " << programName << " " << doc << std::endl;
}

#endif /* SPLIT_CLI */
